### PR TITLE
separate out observability logic

### PR DESCRIPTION
**Pull request by samuelcolvin**
should improve performance a tiny bit, and also will make it easier to replace `last_messages`.

### CODE REVIEW

File: pydantic_ai_slim/pydantic_ai/tools.py [hunk 1]
```diff
@@ -37,14 +47,20 @@ class RunContext(Generic[AgentDeps]):
...
     model: models.Model
     """The model used in this run."""
+    prompt: str
+    """The original user prompt passed to the run."""
+    model_selection_mode: ModelSelectionMode
+    """Dependencies for the agent."""
+    retry: int = 0
+    """Number of retries so far."""
+    run_step: int = 0
+    """The current step in the run."""
```
Comment (sydney-runkle) [inline pydantic_ai_slim/pydantic_ai/tools.py:63]: Awesome to add these, but can we do it in a different PR that focuses just on this enhancement?

File: pydantic_ai_slim/pydantic_ai/tools.py [hunk 2]
```diff
@@ -29,6 +30,15 @@
...
 """Type variable for agent dependencies."""
+ModelSelectionMode = Literal['from-agent', 'custom', 'override']
+"""How the [model][pydantic_ai.models.Model] was selected for the run.
+
+Meanings:
+
+- `'from-agent'`: The model set on the agent was used
+- `'custom'`: The model was set via the `model` kwarg, e.g. on [`run`][pydantic_ai.Agent.run]
+- `'override'`: The model was set by the [`override`][pydantic_ai.Agent.override] function.
+"""
```
Comment (sydney-runkle) [inline pydantic_ai_slim/pydantic_ai/tools.py:41]: Bikeshedding a bit here, could we be even more explicit? 

Like `agent`, `run-kwarg`, and `override`? Feel free to take or leave these particular suggestions. 

I'm hesitant to use "from" with agent, but not any of the others, and "custom" seems not specific enough...

File: pydantic_ai_slim/pydantic_ai/agent.py [hunk 3]
```diff
@@ -219,20 +218,15 @@ async def run(
         """
         if infer_name and self.name is None:
             self._infer_name(inspect.currentframe())
-        model_used, mode_selection = await self._get_model(model)
+        model_used, model_selection_mode = await self._get_model(model)
 
         deps = self._get_deps(deps)
         new_message_index = len(message_history) if message_history else 0
+        run_context = RunContext(deps, [], None, model_used, user_prompt, model_selection_mode)
+        # noinspection PyTypeChecker
```
Comment (sydney-runkle) [inline pydantic_ai_slim/pydantic_ai/agent.py:226]: Why do we need this here?

File: pydantic_ai_slim/pydantic_ai/observe.py
Comment (sydney-runkle) [inline pydantic_ai_slim/pydantic_ai/observe.py:1]: Maybe we could name this `observability`?

**Review by sydney-runkle — COMMENTED** [review]
Seems like a good idea to abstract away some of the observability logic. It's a little bit unfortunate that we have to "pass around" this observer through function calls, but I don't have a better proposal at the moment.

I'd be in favor of moving the changes to `RunContext` into a different PR.

Would love to have a quick call (maybe Monday) to ask a few more questions here just to improve my understanding - motivation for this abstraction pattern, etc.

**Comment by samuelcolvin** [comment]
Overall I don't think this is a good idea, but there are a few good bits we should take out into a new PR.

**Comment by samuelcolvin** [comment]
replaced by #570.
